### ISSUE TITLE
[google_maps_flutter] Should have a callback when map is ready.

### ISSUE THREAD

**Issue by cyanglaz** [flutter/flutter#54758]
Expose below platform method to dart API:

`mapViewDidFinishTileRendering`: on iOS.
`onMapLoaded`: on Android.

**Comment by imeDevelopers** [comment]
It exist as onMapCreated ..

**Comment by cyanglaz** [comment]
@imeDevelopers  ASAIK, `onMapCreated` is called when platform view is ready; not necessary when the map is rendered.

**Comment by imeDevelopers** [comment]
Rendring map are passed on internet .. its not really important idea , because the user see the map directly and you can see all objects even if map not renderd yet like markers , you can use onCameraIdle to change something , its my opinion

**Comment by cyanglaz** [comment]
@imeDevelopers Interesting point about `onCameraIdle`, thanks!

**Comment by Proberts** [comment]
@cyanglaz were you able to simulate a render-complete callback with onCameraIdle?  The map widget in my app has 500ms-1200ms of draw time when navigating back to the map view from another route. I'm trying to cover the redraw using takeSnapshot and Stack. The best I've come up with is using a Future.delayed() in onCameraIdle() to remove the cached image and expose the completed map. However, the draw time worst case range is long enough the user may try to interact with the image. I'm passing through gestures but it pops if the user tries to scroll.

**Comment by cyanglaz** [comment]
@Proberts `onCameraIdle` seems working fine for me. 
Are you talking about the duration from opening a map view to the content of the map view is drawn? It might take time based on the network AFAIK

**Comment by mo-vericatch** [comment]
A typical scenario for this is to center the camera around a marker "once" when the map is finished rendering. with `onCameraIdle` you will need a flag to set to false to prevent it from re-centering again when the map is idle again after dragging.

I think we should have a callback when the rendering is finished for 1 time calls.

**Comment by moga111** [comment]
A typical scenario for this is to center the camera around a marker "once" when the map is finished rendering. with `onCameraIdle` you will need a flag to set to false to prevent it from re-centering again when the map is idle again after dragging.

I think we should have a callback when the rendering is finished for 1 time calls.

**Comment by balaji101010** [comment]
I have a scenario .
I need to take a snapshot when the google map is loaded. 
I take snapshot inside the `onMapCreated()` but it takes snapshot while the google map is rendering on some devices . I get blurry unrendered snapshot.

so I wait for 1 secs then I take the screenshot!

```
/// state
class GoogleMapWidgetState extends State<GoogleMapWidget> {
  final Completer<GoogleMapController> _mapController =
      Completer<GoogleMapController>();

  ///
  Uint8List? imageBytes;

  @override
  Widget build(BuildContext context) {
    return imageBytes != null
        ? Image.memory(imageBytes!, fit: BoxFit.cover)
        : GoogleMap(
            onMapCreated: (GoogleMapController controller) {
              _mapController.complete(controller);
              takeSnapShot();
            },
            initialCameraPosition: CameraPosition(
              target: widget.latLng,
              zoom: widget.zoomLevel,
            ),
            markers: widget.disruptionMarkers,
            compassEnabled: false,
            indoorViewEnabled: false,
            myLocationEnabled: false,
            myLocationButtonEnabled: false,
            zoomControlsEnabled: false,
            zoomGesturesEnabled: false,
            tiltGesturesEnabled: false,
            scrollGesturesEnabled: false,
            rotateGesturesEnabled: false,
          );
  }

  ///
  void takeSnapShot() async {
    GoogleMapController controller = await _mapController.future;
    Future<void>.delayed(const Duration(milliseconds: 1000), () async {
      imageBytes = await controller.takeSnapshot();
      setState(() {});
    });
  }
}
```
To avoid this,
 please provide the `callback` when the map finished the rendered!!!

**Comment by rgplvr** [comment]
This is an issue when we try to load the data based on the visible region, if we try using mapcreated the getVisible region is giving 0.0 for lat and lon.

**Comment by mossmana** [comment]
@stuartmorgan Any chance we can up the priority on this issue? Currently there's a workaround in the google maps  for Android integration tests that adds up to 5 unnecessary seconds delay each time the tests are run. i.e. `await Future<void>.delayed(const Duration(seconds: 1))` is called 5 times. This proposal could potentially speed the tests up.

**Comment by stuartmorgan** [comment]
> Currently there's a workaround in the google maps for Android integration tests that adds up to 5 unnecessary seconds delay each time the tests are run. i.e. `await Future<void>.delayed(const Duration(seconds: 1))` is called 5 times. This proposal could potentially speed the tests up. It may also help alleviate some of the flakiness if 1 second is not quite long enough.

Is the test waiting for a specific value? If so, we can address both of those issues by just putting the test in a poll loop with short interval and longer max. E.g., check if we have the expected values every 0.1 seconds for up to 5 or 10 seconds.

**Comment by mossmana** [comment]
Maybe @cyanglaz has more context, but from what I can tell, polling won't be sufficient in this particular case, as the visible region needs to be checked only when the map has completely finished rendering (i.e. fully visible and not blurry to the human eye). The only way to guarantee the correct value is to have access to the callback. e.g. https://github.com/flutter/packages/blob/33dacd6d67e62fee910be02eac04ec5402bf270b/packages/google_maps_flutter/google_maps_flutter_android/example/integration_test/google_maps_tests.dart#L907

**Comment by cyanglaz** [comment]
> Maybe @cyanglaz has more context, but from what I can tell, polling won't be sufficient in this particular case, as the visible region needs to be checked only when the map has completely finished rendering (i.e. fully visible and not blurry to the human eye). The only way to guarantee the correct value is to have access to the callback. e.g. https://github.com/flutter/packages/blob/33dacd6d67e62fee910be02eac04ec5402bf270b/packages/google_maps_flutter/google_maps_flutter_android/example/integration_test/google_maps_tests.dart#L907

Yeah this most of the tests, polling would work. This particular one seems to be waiting for the "visibleRegion" to be "stabilized" so it is not checking a particular value. I think the implementing the callback in this issue would help. However, we can def implement polling for other tests to speed the tests up and deflake them.
